### DISCUSSION TITLE
Information about testing

### DISCUSSION THREAD

**Discussion by mirpedrol** [tconbeer/harlequin#551] [Q&A, Troubleshooting, Support] [answered]
Hi @tconbeer!

I am working on adding a TUI to our package [nf-core/tools](https://github.com/nf-core/tools), currently implementing the snapshot tests.

I see your tests are asyncio events and then you are able to generate a snapshot with `[app_snapshot()](https://github.com/tconbeer/harlequin/blob/974d8f7b922aa1b4b4b4d95f4309c67a7b060121/tests/functional_tests/test_app.py#L62)`. Could you point me to how to use this same approach?
When I try to add it to my tests I get an error `fixture 'app_snapshot' not found`

Thank you very much for your help!

**Comment by tconbeer** [comment, marked as answer]
`app_snapshot` is a pytest fixture that is defined by the `pytest-textual-snapshot` test dependency in Harlequin. I use [my fork](https://github.com/tconbeer/pytest-textual-snapshot) of [Textualize's package](https://github.com/Textualize/pytest-textual-snapshot). You need to install that package as a dependency before using the fixture; I don't recommend installing from my repo without forking it first, since I might change that at any time for Harlequin's purposes and break your stuff.

Basically, it uses Textual's built-in screenshotting capabilities and stores the SVG screenshot as a data snapshot with [Syrupy](https://github.com/tophat/syrupy). If the SVG changes between test runs, the test fails.

Textualize's package uses a fixture called `snap_compare`, which takes the path to an App file and a sequence of keystrokes. My fork adds the `app_snapshot` fixture, which you can call from inside a test script that uses the [Pilot](https://textual.textualize.io/api/pilot/) to drive the tests, which allows for multiple assertions per test setup.

These tests are easy to get started with, but frankly a pain to maintain. They are super brittle (changing a single pixel on the screen will fail the test -- sometimes the changes aren't even visible once the browser renders the SVG, since they're to hidden elements, etc.). They also execute pretty slowly. The tradeoffs are worth it to me, because I really care about quality, but they significantly slow down development.

**Reply by mirpedrol** [reply]
Thank you @tconbeer, this was very useful
